File: Library/Windows/HardwareLib/HardwareDaemon/Proto/CrossApi.cs
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;
using HardwareDaemon.Hardware;
using Proto.Generated.PCrossApi;
using static HardwareDaemon.State.ServiceState;

namespace HardwareDaemon.Proto;

public class CrossApi : PCrossApi.PCrossApiBase
{
    private readonly ILogger<CrossApi> _logger;

    public CrossApi(ILogger<CrossApi> logger)
    {
        _logger = logger;
    }


    public override Task<POk> POpen(Empty request, ServerCallContext context)
    {
        Console.WriteLine("[SERVICE] open");

        IsOpen = true;

        var response = new POk
        {
            PIsSuccess = true
        };

        return Task.FromResult(response);
    }

    private static PHardwareList CreatePHardwareList(IEnumerable<BaseHardware> list)
    {
        var hardwareList = new PHardwareList();
        foreach (var hardware in list)
        {
            var pHardware = new PHardware
            {
                PName = hardware.Name,
                PId = hardware.Id
            };
            hardwareList.PHardwares.Add(pHardware);
        }

        return hardwareList;
    }

    public override Task<PHardwareList> PGetHardware(PHardwareTypeMessage request, ServerCallContext context)
    {
        Console.WriteLine("[SERVICE] getHardware: " + request.PType);

        var hardwareList = request.PType switch
        {
            PHardwareType.Control => CreatePHardwareList(State.HControls.Values),
            PHardwareType.Temp => CreatePHardwareList(State.HTemps.Values),
            PHardwareType.Fan => CreatePHardwareList(State.HFans.Values),
            _ => throw new ProtoException("unknown pHardware type")
        };

        Console.WriteLine("[SERVICE] " + hardwareList.PHardwares.Count);

        return Task.FromResult(hardwareList);
    }


    private static PUpdateList CreatePUpdate(PHardwareType type, IEnumerable<BaseHardware> list)
    {
        var updateList = new PUpdateList
        {
            PType = type
        };

        foreach (var hardware in list)
        {
            var update = new PUpdate
            {
                PIndex = hardware.Index,
                PValue = hardware.Value
            };
            updateList.PUpdates.Add(update);
        }

        return updateList;
    }

    public override async Task PStartStream(Empty request, IServerStreamWriter<PUpdateList> responseStream,
        ServerCallContext context)
    {
        while (!context.CancellationToken.IsCancellationRequested && IsOpen)
        {
            HardwareManager.Update();
            Update.UpdateAllSensors();

            var updateList = CreatePUpdate(PHardwareType.Control, State.HControls.Values);

            await responseStream.WriteAsync(updateList, context.CancellationToken);

            updateList = CreatePUpdate(PHardwareType.Temp, State.HTemps.Values);
            await responseStream.WriteAsync(updateList, context.CancellationToken);

            updateList = CreatePUpdate(PHardwareType.Fan, State.HFans.Values);
            await responseStream.WriteAsync(updateList, context.CancellationToken);

            await Task.Delay(State.Settings.UpdateDelay * 1000, context.CancellationToken);
        }
    }


    public override Task<Empty> PCloseStream(Empty request, ServerCallContext context)
    {
        Console.WriteLine("[SERVICE] close");
        IsOpen = false;

        return Task.FromResult(new Empty());
    }

    public override Task<POk> PSettingsAndConfChange(Empty request, ServerCallContext context)
    {
        Console.WriteLine("[SERVICE] confChange");

        SettingsAndConfHasChange = true;

        var response = new POk
        {
            PIsSuccess = true
        };

        return Task.FromResult(response);
    }

    public override Task<POk> PSettingsChange(Empty request, ServerCallContext context)
    {
        Console.WriteLine("[SERVICE] settings Change");

        SettingsHasChange = true;

        var response = new POk
        {
            PIsSuccess = true
        };

        return Task.FromResult(response);
    }
}

File: FanControl/app/src/jvmMain/kotlin/proto/CrossApi.kt
package proto

import FState
import FState.hControls
import FState.hFans
import FState.hTemps
import ServiceState
import com.google.protobuf.Empty
import io.grpc.ManagedChannel
import io.grpc.StatusException
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.runBlocking
import model.hardware.HControl
import model.hardware.HFan
import model.hardware.HTemp
import proto.generated.pCrossApi.PCrossApiGrpcKt
import proto.generated.pCrossApi.PHardwareType
import proto.generated.pCrossApi.PUpdateList
import proto.generated.pCrossApi.pHardwareTypeMessage
import java.io.Closeable
import java.util.concurrent.TimeUnit

class CrossApi(
    private val channel: ManagedChannel,
) : Closeable {
    private val stub: PCrossApiGrpcKt.PCrossApiCoroutineStub = PCrossApiGrpcKt.PCrossApiCoroutineStub(channel)


    // don't always relevant
    private fun isActive(): Boolean {
        val isActive =
            FState.service.status.value == ServiceState.Status.OPEN && !channel.isShutdown && !channel.isTerminated

        if (!isActive) {
            println("service is not active")
            FState.service.setErrorStatus()
            return false
        }
        return true
    }

    suspend fun open(): Boolean {
        if (channel.isShutdown || channel.isTerminated) {
            println("api open: failed, channel is not valid")
            return false
        }

        return try {
            val res = stub.pOpen(Empty.getDefaultInstance())

            when (res.pIsSuccess) {
                true -> {
                    println("open success")
                    true
                }

                false -> {
                    println("open service returned false")
                    false
                }
            }
        } catch (e: StatusException) {
            println("Error: can't open, service status = ${e.status.code}")
            false
        } catch (e: Exception) {
            e.printStackTrace()
            FState.service.setErrorStatus()
            false
        }
    }

    suspend fun getHardware(): Boolean {
        if (!isActive()) return false

        try {
            val pControls = stub.pGetHardware(pHardwareTypeMessage { pType = PHardwareType.CONTROL })
            hControls.apply {
                clear()
                addAll(pControls.pHardwaresList.map {
                    HControl(
                        name = it.pName,
                        id = it.pId
                    )
                })
            }

            val pTemps = stub.pGetHardware(pHardwareTypeMessage { pType = PHardwareType.TEMP })
            hTemps.apply {
                clear()
                addAll(pTemps.pHardwaresList.map {
                    HTemp(
                        name = it.pName,
                        id = it.pId
                    )
                })
            }

            val pFans = stub.pGetHardware(pHardwareTypeMessage { pType = PHardwareType.FAN })
            hFans.apply {
                clear()
                addAll(pFans.pHardwaresList.map {
                    HFan(
                        name = it.pName,
                        id = it.pId
                    )
                })
            }
            println("getHardware success")
            return true

        } catch (e: Exception) {
            e.printStackTrace()
            FState.service.setErrorStatus()
            return false
        }
    }


    suspend fun startUpdate() {
        if (!isActive()) return

        val stream: Flow<PUpdateList> = stub.pStartStream(Empty.getDefaultInstance())

        println("streaming has started")

        try {
            stream.collect { updateList ->
                when (updateList.pType) {
                    PHardwareType.CONTROL -> {
                        updateList.pUpdatesList.forEach {
                            hControls[it.pIndex].value.value = it.pValue
                        }
                    }

                    PHardwareType.TEMP -> {
                        updateList.pUpdatesList.forEach {
                            hTemps[it.pIndex].value.value = it.pValue
                        }
                    }

                    PHardwareType.FAN -> {
                        updateList.pUpdatesList.forEach {
                            hFans[it.pIndex].value.value = it.pValue
                        }
                    }

                    else -> throw ProtoException("unknown control type")
                }
            }
        } catch (e: Exception) {
            if (!shouldClose) {
                shouldClose = false
                e.printStackTrace()
            }
            println("ERROR: stream update has stop")
            FState.service.setErrorStatus()
        }

    }

    private var shouldClose = false
    override fun close() {
        shouldClose = true

        try {
            if (isActive()) {
                runBlocking {
                    stub.pCloseStream(Empty.getDefaultInstance())
                }
            }

        } catch (e: Exception) {
            println("close has failed")
        }
        try {
            if (!channel.isShutdown) {
                channel.shutdown().awaitTermination(5, TimeUnit.SECONDS)
            }
        } catch (e: Exception) {
            println("shutdown has failed")
        }

        println("close api finished")
    }


    suspend fun settingsAndConfChange() {
        if (!isActive()) return

        try {
            val res = stub.pSettingsAndConfChange(Empty.getDefaultInstance())

            when (res.pIsSuccess) {
                true -> println("confChange success")
                false -> println("confChange failed")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            FState.service.setErrorStatus()
        }


    }

    suspend fun settingsChange() {
        if (!isActive()) return

        try {
            val res = stub.pSettingsChange(Empty.getDefaultInstance())

            when (res.pIsSuccess) {
                true -> println("settingsChange success")
                false -> println("settingsChange failed")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            FState.service.setErrorStatus()
        }

    }
}